cli/src/cli_util.rs

```diff
@@ -3252,70 +3250,88 @@ fn resolve_default_command(
                     ui.hint_no_heading(),
                     "Run `jj config set --user ui.default-command log` to disable this message."
                 )?;
+
+                vec!["log".to_string()]
             }
-            let default_command = args.unwrap_or_else(|| vec!["log".to_string()]);
+        };
 
-            // Insert the default command directly after the path to the binary.
-            string_args.splice(1..1, default_command);
-        }
+        // Insert the default command directly after the path to the binary.
+        cmdline.splice(1..1, default_args).for_each(std::mem::drop);
     }
-    Ok(string_args)
+
+    Ok(())
 }
 
-fn resolve_aliases(
+/// Expand any aliases in the supplied command line.
+fn expand_cmdline_aliases(
     ui: &Ui,
     config: &StackedConfig,
     app: &Command,
-    mut string_args: Vec<String>,
-) -> Result<Vec<String>, CommandError> {
-    let defined_aliases: HashSet<_> = config.table_keys("aliases").collect();
-    let mut resolved_aliases = HashSet::new();
+    cmdline: &mut Vec<String>,
+) -> Result<(), CommandError> {
+    // Collect all the commands defined in clap.
     let mut real_commands = HashSet::new();
     for command in app.get_subcommands() {
         real_commands.insert(command.get_name());
         for alias in command.get_all_aliases() {
             real_commands.insert(alias);
         }
     }
-    for alias in defined_aliases.intersection(&real_commands).sorted() {
+
+    // Collect all the aliases defined by the user.
+    let defined_aliases: HashMap<String, Vec<String>> = config.get("aliases")?;
```

Comment:
> The only significant one I'm aware of is that jj will now always complain about
malformed aliases, instead of waiting for first use.

It might be okay to warn all bad aliases, but I think it's better to expand valid aliases instead of rejecting everything.

Reply:
On second thought, keeping the original behaviour is probably better.